rfc/router-storage.md

```diff
@@ -0,0 +1,767 @@
+
+The blob router contains a state of storage nodes available.
+
+```ts
+type StorageNode = {
+  did: ProviderDID
+  w3upEndpoint: URL // a UCAN invocation endpoint providing at minimum `service/blob/allocate` and `service/blob/accept`
+  storageProperties?: StorageProperties
+}
+
+type StorageNodes = {
+  [ProviderDID]: StorageNode
+}
+
+// intentionally left blank for now to expand over time
+type StorageProperties = {
+}
+```
+
+The blob router will probably also need a state of where things have been placed at and what stage they're at:
+
+```ts
+type BlobAllocation = {
+  did: ProviderDid
+  blob: Blob
+}
+
+type BlobAllocations { [digest]: BlobAllocation }
+
+```
+
+For now, we don't define any protocols for adding and removing providers to a router. Moreover, expecting what we track about nodes will evolve over time, for now we are not attempting to define specific storage properties for now
+
+### `space/content/add/blob` execution flow
+
+#### `space/content/add/blob` invoked on space:
+
+```json
+{ // "/": "bafy..add"
+  "cmd": "/space/content/add/blob",
+  "sub": "did:key:zAlice",
+  "iss": "did:key:zAlice",
+  "aud": "did:web:web3.storage",
+  "args": {
+    "blob": {
+      // multihash of the blob as byte array
+      "digest": { "/": { "bytes": "mEi...sfKg" } },
+      // size of the blob in bytes
+      "size": 2_097_152,
+    },
+    "storagePreferences": {
+
+    }
+  }
+}
+```
+
+Legacy invocation of `blob/add`:
+
+```json
+{ // "/": "bafy..add"
+  
+  "iss": "did:key:zAlice",
+  "aud": "did:web:web3.storage",
+  "att": [
+    {
+      "cmd": "/space/content/add/blob",
+      "with": "did:key:zAlice",
+      "nb": {
+        "blob": {
+          // multihash of the blob as byte array
+          "digest": { "/": { "bytes": "mEi...sfKg" } },
+          // size of the blob in bytes
+          "size": 2_097_152,
+        },
+        "storagePreferences": {
+        }
+      }
+    }
+  ]
+}
+```
+
+Receipt:
+
+```json
+{ // "/": "bafy..work",
+  "iss": "did:web:web3.storage",
+  "aud": "did:key:zAlice",
+  "cmd": "/ucan/assert",
+  "sub": "did:web:web3.storage",
+  "args": {
+    "assert": [
+      // refers to the invocation from the example
+      { "/": "bafy..add" },
+      // refers to the receipt corresponding to the above invocation
+      {
+        "out": {
+          "ok": {
+            // result of the add is the content (location) commitment
+            // that is produced as result of "bafy..accept"
+            "site": {
+              "ucan/await": [
+                ".out.ok.site",
+                { "/": "bafy...accept" }
+              ]
+            }
+          }
+        },
+        // Previously `next` was known as `fx` instead, which is
+        // set of tasks to be scheduled.
+        "next": [
+          // 1. System attempts to allocate memory in user space for the blob.
+          { // "/": "bafy...alloc",
+            "cmd": "/service/blob/allocate",
+            "sub": "did:web:web3.storage",
```

Comment:
ok, sorry, stop rewind need to check for understanding here.

The fundamental problem is that while we are delegating a /service/blob/allocate to storage node, we may have to try multiple nodes first in order to succeed. The proposed architecture treats that negotiation as an asynchronous process, with a higher level "wrapper" version of /service/blob/allocate running on the router, and this being returned in the invocation chain rather than an invocation on a specific node.

An alternate approach is to do that entire negotiation synchronously, and simply discard any failures in the invocation chain, returning the one that succeeded. I think this is much simple, but it does hide the failed invocations and negotiation from the user. More importantly, it could take a while potentially, depending on node acceptance rate and time to respond.

Given your suggestion here, I believe that you are assuming an approach closer this alternative -- can I confirm that is what you are thinking.

Reply:
> The fundamental problem is that while we are delegating a /service/blob/allocate to storage node

Other way round storage node delegates ability to invoke that to the router, that way router can allocate space on any registered storage node as needed.

> Given your suggestion here, I believe that you are assuming an approach closer this alternative -- can I confirm that is what you are thinking.

I am assuming async not the sync, not sure which one you were referring to as an alternative. That said I think both options should be possible if we get a delegation from the storage node to perform an allocation, whether we expose decision making details from router to user or not is something that can be decided separately.